README.md
rclpy_message_converter
=======================

rclpy_message_converter is a lightweight ROS2 package and Python library to
convert from Python dictionaries and JSON messages to rclpy messages, and vice
versa.

ROS 1 and ROS 2 branches
------------------------

ROS 1 users should use the `master` branch. ROS 2 users should use the appropriate
branch for their distro (`foxy`/`galactic`/`humble`/`rolling`/...).

Note to ROS 2 users
-------------------

Most ROS 2 users probably *do not need* this package. This package was originally
developed for ROS 1. In ROS 2, most of the functionality is already built in.
Specifically, check out

* [rosidl_runtime_py.set_message.set_message_fields()](https://github.com/ros2/rosidl_runtime_py/blob/1979f566c3b446ddbc5c3fb6896e1f03ccbc6a27/rosidl_runtime_py/set_message.py#L28-L37)
  for converting a Python dictionary to a ROS message, and
* [rosidl_runtime_py.convert.message_to_orderreddict](https://github.com/ros2/rosidl_runtime_py/blob/1979f566c3b446ddbc5c3fb6896e1f03ccbc6a27/rosidl_runtime_py/convert.py#L159-L176)
  for converting a ROS message to a Python dictionary.

The `rclpy_message_converter` mainly adds the functionality to encode all
variable-size `uint8[]` or fixed-size `uint8[n]` fields using Base64 encoding
when the parameter `base64_encoding` is set.
This saves a lot of space when converting large messages
(such as `sensor_msgs/Image`) to JSON format. Also, encoding/decoding of Base64
strings is required by some packages (such as [mir_robot](https://github.com/dfki-ric/mir_robot))
that were originally developed in ROS1.

Usage
-----

Convert a dictionary to a ROS message

```python
from rclpy_message_converter import message_converter
from std_msgs.msg import String
dictionary = { 'data': 'Howdy' }
message = message_converter.convert_dictionary_to_ros_message('std_msgs/msg/String', dictionary)
```

Convert a ROS message to a dictionary

```python
from rclpy_message_converter import message_converter
from std_msgs.msg import String
message = String(data = 'Howdy')
dictionary = message_converter.convert_ros_message_to_dictionary(message)
```

Convert JSON to a ROS message

```python
from rclpy_message_converter import json_message_converter
from std_msgs.msg import String
json_str = '{"data": "Hello"}'
message = json_message_converter.convert_json_to_ros_message('std_msgs/msg/String', json_str)
```

Convert a ROS message to JSON

```python
from rclpy_message_converter import json_message_converter
from std_msgs.msg import String
message = String(data = 'Hello')
json_str = json_message_converter.convert_ros_message_to_json(message)
```

Test
----

To run the tests:

```bash
colcon test
```

pre-commit Formatting Checks
----------------------------

This repo has a [pre-commit](https://pre-commit.com/) check that runs in CI.
You can use this locally and set it up to run automatically before you commit
something. To install, use pip:

```bash
pip3 install --user pre-commit
```

To run over all the files in the repo manually:

```bash
pre-commit run -a
```

To run pre-commit automatically before committing in the local repo, install the git hooks:

```bash
pre-commit install
```


License
-------

This project is released under the Apache 2.0 license.


CHANGELOG.rst
Change Log
==========

0.5.9 (2022-09-12)
------------------
* Fix flake8 errors
* Re-format code using black
* package.xml: Add missing build_export_depend
* Fix EOF and trailing whitespace
* Add pre-commit config
* README: Add note about branches
* pass down log_level to helper functions (`#60 <https://github.com/uos/rospy_message_converter/issues/60>`_)
* Declare file encoding
  This is necessary on ROS Melodic (Python2), because I have added a
  non-ASCII character (u umlaut) in my last commit.
* Add LICENSE file and license headers
* Contributors: Martin Günther, Yuri Rocha

0.5.8 (2022-03-03)
------------------
* add option to change log level (`#58 <https://github.com/uos/rospy_message_converter/issues/58>`_)
* Contributors: Yuri Rocha

0.5.7 (2021-09-08)
------------------
* Handle None values in dictionary
* Add tests for None
* Dockerfile-kinetic: Add --include-eol-distros
* Contributors: Martin Günther

0.5.6 (2021-03-01)
------------------
* Propagate strict_mode, check_missing_fields in _convert_to_ros_type
  Previously, _convert_to_ros_type dropped strict_mode and
  check_missing_fields in nested messages.
* Add NestedUint8ArrayTestService tests
* propagate check_types in _convert_to_ros_type (`#51 <https://github.com/uos/rospy_message_converter/issues/51>`_)
  Co-authored-by: Martin Günther <martin.guenther@dfki.de>
* Fix binary_array_as_bytes=False with nested msgs
* Add param binary_array_as_bytes
  Closes `#45 <https://github.com/uos/rospy_message_converter/issues/45>`_.
* Contributors: Marc Bosch-Jorge, Martin Günther, Otacon5555

0.5.5 (2020-11-09)
------------------
* Decode strings from ROS messages as UTF8
  This makes the python2 behavior equal to python3.
* python3 only: Validate base64 strings
* Add bytes to python3 string types
  This means that `bytes` will now also be base64-decoded, which fixes the following tests on python3:
  * test_dictionary_with_uint8_array_bytes
  * test_dictionary_with_uint8_array_bytes_unencoded
  * test_dictionary_with_3uint8_array_bytes
  On python2, `bytes` is just an alias for `str`, which is why it worked
  without this.
* Fix and add tests
* Contributors: Martin Günther

0.5.4 (2020-10-13)
------------------
* Avoid numpy dependency
* Contributors: Martin Günther, betaboon

0.5.3 (2020-08-20)
------------------
* Add check_types parameter to convert_dictionary_to_ros_message (`#42 <https://github.com/uos/rospy_message_converter/issues/42>`_)
* Allow numpy numeric types in numeric fields  (`#41 <https://github.com/uos/rospy_message_converter/issues/41>`_)
  Fixes `#39 <https://github.com/uos/rospy_message_converter/issues/39>`_.
* perf: Remove remaining regexes
  This is only a small speedup of about 1.03x.
* perf: Avoid regex in _is_field_type_a_primitive_array
  This makes the function almost 3x faster.
* perf: Reorder type checks
  Perform the cheaper checks first. This results in a speedup of about
  1.2x.
* perf: Avoid regex in is_ros_binary_type
  This makes is_ros_binary_type almost 5x faster and as a result the whole
  convert_ros_message_to_dictionary function almost 2x faster.
* Compare types, not type names; improve error message
  Old error message:
  TypeError: Wrong type: '1.0' must be float64
  New error message:
  TypeError: Field 'x' has wrong type <type 'numpy.float64'> (valid types: [<type 'int'>, <type 'float'>])
* Remove unused python_to_ros_type_map
* added test for convert_dictionary_to_ros_message with int8 array
* python 3 fix for _convert_to_ros_binary
* Contributors: Martin Günther, Steffen Rühl

0.5.2 (2020-07-09)
------------------
* Check for wrong field types when converting from dict to ros msg
* Check for missing fields when converting from dict to ros msg
* Contributors: Martin Günther, alecarnevale

0.5.1 (2020-05-25)
------------------
* Initial release into Noetic
* Decode base64-encoded byte arrays as unicode
* Make tests compatible with python3
* add check for python3 str serializing `#33 <https://github.com/uos/rospy_message_converter/issues/33>`_ (`#34 <https://github.com/uos/rospy_message_converter/issues/34>`_)
* efficient conversion of primitive array to ros type (`#31 <https://github.com/uos/rospy_message_converter/issues/31>`_)
* efficient conversion of primitive array
* removed unused _convert_from_ros_primitive
* optionally ignore extra fields when deserializing (`#29 <https://github.com/uos/rospy_message_converter/issues/29>`_)
* Remove EOL distros indigo + lunar from CI
* travis CI: Use matrix to split ROS distros
* Update README (convert to md, add build status)
* Contributors: Martin Günther, George Hartt, Jannik Abbenseth, Omri Rozenzaft

0.5.0 (2019-01-17)
------------------
* Initial release into Lunar and Melodic
* Remove support for Jade (EOL)
* Change maintainer from Brandon Alexander to Martin Günther
* Move repo from baalexander to uos
* Add serialize_deserialize to unit tests, fix incorrect tests caught by this
* Remove dependency on ROS master in tests; all tests are now unit
  tests  (`#18 <https://github.com/uos/rospy_message_converter/issues/18>`_)
* Add service request/response support (`#17 <https://github.com/uos/rospy_message_converter/issues/17>`_)
* Fix fixed-size uint8 array conversion failure (`#15 <https://github.com/uos/rospy_message_converter/issues/15>`_)
* Fix unicode handling in string fields (`#13 <https://github.com/uos/rospy_message_converter/issues/13>`_)
* Enable testing only if CATKIN_ENABLE_TESTING is set (`#9 <https://github.com/uos/rospy_message_converter/issues/9>`_)
* Contributors: Martin Günther, Brandon Alexander, George Laurent, Jean-Baptiste Doyon, Viktor Schlegel, Rein Appeldoorn, Will Baker, neka-nat

0.4.0 (2015-12-13)
------------------
* Adds support for ROS Jade
* Removes support for ROS Groovy and Hydro (EOL)
* Uses single branch for all ROS versions
* Docker support for local development and Travis CI

0.3.0 (2014-06-03)
------------------
* Adds support for ROS Indigo

0.2.0 (2013-07-15)
------------------

test/test_json_message_converter.py
# -*- coding: utf-8 -*-
#
# Software License Agreement (BSD License)
#
# Copyright (c) 2019-2022, Martin Günther (DFKI GmbH) and others
# Copyright (c) 2013-2016, Brandon Alexander
#
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of this project nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import unittest
import rospy
from rospy_message_converter import json_message_converter

class TestJsonMessageConverter(unittest.TestCase):

    def test_ros_message_with_string(self):
        from std_msgs.msg import String
        expected_json = '{"data": "Hello"}'
        message = String(data = 'Hello')
        message = serialize_deserialize(message)
        returned_json = json_message_converter.convert_ros_message_to_json(message)
        self.assertEqual(returned_json, expected_json)

    def test_ros_message_with_string_unicode(self):
        from std_msgs.msg import String
        expected_json = '{"data": "Hello \\u00dcnicode"}'
        message = String(data = u'Hello \u00dcnicode')
        message = serialize_deserialize(message)
        returned_json = json_message_converter.convert_ros_message_to_json(message)
        self.assertEqual(returned_json, expected_json)

    def test_ros_message_with_header(self):
        from std_msgs.msg import Header
        from time import time
        now_time = rospy.Time(time())
        expected_json1 = '{{"stamp": {{"secs": {0}, "nsecs": {1}}}, "frame_id": "my_frame", "seq": 3}}'\
            .format(now_time.secs, now_time.nsecs)
        expected_json2 = '{{"seq": 3, "stamp": {{"secs": {0}, "nsecs": {1}}}, "frame_id": "my_frame"}}'\
            .format(now_time.secs, now_time.nsecs)
        expected_json3 = '{{"frame_id": "my_frame", "seq": 3, "stamp": {{"secs": {0}, "nsecs": {1}}}}}'\
            .format(now_time.secs, now_time.nsecs)
        message = Header(stamp = now_time, frame_id = 'my_frame', seq = 3)
        message = serialize_deserialize(message)
        returned_json = json_message_converter.convert_ros_message_to_json(message)
        self.assertTrue(returned_json == expected_json1 or
                        returned_json == expected_json2 or
                        returned_json == expected_json3)

    def test_ros_message_with_uint8_array(self):
        from rospy_message_converter.msg import Uint8ArrayTestMessage
        input_data = [97, 98, 99, 100]
        expected_json = '{"data": "YWJjZA=="}'  # base64.b64encode("abcd") is "YWJjZA=="
        message = Uint8ArrayTestMessage(data=input_data)
        message = serialize_deserialize(message)
        returned_json = json_message_converter.convert_ros_message_to_json(message)
        self.assertEqual(returned_json, expected_json)

    def test_ros_message_with_3uint8_array(self):
        from rospy_message_converter.msg import Uint8Array3TestMessage
        input_data = [97, 98, 99]
        expected_json = '{"data": "YWJj"}'  # base64.b64encode("abc") is "YWJj"
        message = Uint8Array3TestMessage(data=input_data)
        message = serialize_deserialize(message)
        returned_json = json_message_converter.convert_ros_message_to_json(message)
        self.assertEqual(returned_json, expected_json)

    def test_json_with_string(self):
        from std_msgs.msg import String
        expected_message = String(data = 'Hello')
        json_str = '{"data": "Hello"}'
        message = json_message_converter.convert_json_to_ros_message('std_msgs/String', json_str)
        expected_message = serialize_deserialize(expected_message)
        self.assertEqual(message, expected_message)

    def test_json_with_string_unicode(self):
        from std_msgs.msg import String
        expected_message = String(data = u'Hello \u00dcnicode')
        json_str = '{"data": "Hello \\u00dcnicode"}'
        message = json_message_converter.convert_json_to_ros_message('std_msgs/String', json_str)
        expected_message = serialize_deserialize(expected_message)
        self.assertEqual(message, expected_message)

    def test_json_with_header(self):
        from std_msgs.msg import Header
        from time import time
        now_time = rospy.Time(time())
        expected_message = Header(
            stamp = now_time,
            frame_id = 'my_frame',
            seq = 12
        )
        json_str = '{{"stamp": {{"secs": {0}, "nsecs": {1}}}, "frame_id": "my_frame", "seq": 12}}'\
            .format(now_time.secs, now_time.nsecs)
        message = json_message_converter.convert_json_to_ros_message('std_msgs/Header', json_str)
        expected_message = serialize_deserialize(expected_message)
        self.assertEqual(message, expected_message)

    def test_json_with_string_null(self):
        from std_msgs.msg import String
        expected_message = String(data='')
        json_str = '{"data": null}'
        message = json_message_converter.convert_json_to_ros_message('std_msgs/String', json_str)
        expected_message = serialize_deserialize(expected_message)
        self.assertEqual(message, expected_message)

    def test_json_with_invalid_message_fields(self):
        self.assertRaises(ValueError,
                          json_message_converter.convert_json_to_ros_message,
                          'std_msgs/String',
                          '{"not_data": "Hello"}')


def serialize_deserialize(message):
    """
    Serialize and then deserialize a message. This simulates sending a message
    between ROS nodes and makes sure that the ROS messages being tested are
    actually serializable, and are in the same format as they would be received
    over the network. In rospy, it is possible to assign an illegal data type
    to a message field (for example, `message = String(data=42)`), but trying
    to publish this message will throw `SerializationError: field data must be
    of type str`. This method will expose such bugs.
    """
    from io import BytesIO
    buff = BytesIO()
    message.serialize(buff)
    result = message.__class__()   # create new instance of same class as message
    result.deserialize(buff.getvalue())
    return result


PKG = 'rospy_message_converter'
NAME = 'test_json_message_converter'
if __name__ == '__main__':
    import rosunit
    rosunit.unitrun(PKG, NAME, TestJsonMessageConverter)


rclpy_message_converter/rclpy_message_converter/json_message_converter.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2019-2022, Martin Günther (DFKI GmbH) and others
# Copyright (c) 2013-2016, Brandon Alexander
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json

from rclpy_message_converter import message_converter


def convert_json_to_ros_message(message_type, json_message, strict_mode=True):
    """
    Takes in the message type and a JSON-formatted string and returns a ROS
    message.

    :param message_type: The desired ROS message type of the result
    :type  message_type: str
    :param json_message: A JSON-formatted string
    :type  json_message: str
    :param strict_mode: If strict_mode is set, an exception will be thrown if the json message contains extra fields.
    :type  strict_mode: bool, optional
    :return: A ROS message
    :rtype: class:`genpy.Message`

    Example:
        >>> msg_type = "std_msgs/msg/String"
        >>> json_msg = '{"data": "Hello, Robot"}'
        >>> convert_json_to_ros_message(msg_type, json_msg)
        std_msgs.msg.String(data='Hello, Robot')
    """
    dictionary = json.loads(json_message)
    return message_converter.convert_dictionary_to_ros_message(message_type, dictionary, strict_mode=strict_mode)


def convert_ros_message_to_json(message, base64_encoding=True):
    """
    Takes in a ROS message and returns a JSON-formatted string.

    :param message: A ROS message to convert
    :type  message: class:`genpy.Message`
    :param base64_encoding: If this parameter is true, encode all variable-size `uint8[]` or fixed-size `uint8[n]`
           fields using Base64 encoding. This saves a lot of space when converting large messages
           (such as sensor_msgs/Image) to json format.
           If this parameter is `False`, these fields will be converted to `array.array` (for variable-size fields) or
           `numpy.ndarray` (for fixed-size fields) instead.
    :type  base64_encoding: bool, optional
    :return: A JSON-formatted string
    :rtype:  str

    Example:
        >>> import std_msgs.msg
        >>> ros_message = std_msgs.msg.String(data="Hello, Robot")
        >>> convert_ros_message_to_json(ros_message)
        '{"data": "Hello, Robot"}'
    """
    dictionary = message_converter.convert_ros_message_to_dictionary(message, base64_encoding)
    json_message = json.dumps(dictionary)
    return json_message


if __name__ == "__main__":
    import doctest

    doctest.testmod()


rclpy_message_converter/rclpy_message_converter/message_converter.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2019-2022, Martin Günther (DFKI GmbH) and others
# Copyright (c) 2017-2019 Open Source Robotics Foundation, Inc.
# Copyright (c) 2013-2016, Brandon Alexander
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import array
import base64
from collections import OrderedDict
from typing import Any, Dict

import numpy as np
import rosidl_parser.definition
from rosidl_parser.definition import AbstractNestedType, NamespacedType, Array, UnboundedSequence, BasicType
from rosidl_runtime_py.convert import get_message_slot_types
from rosidl_runtime_py.import_message import import_message_from_namespaced_type
from rosidl_runtime_py.utilities import get_message, get_service


def convert_dictionary_to_ros_message(
    message_type: str,
    dictionary: Dict[str, Any],
    kind: str = 'message',
    strict_mode: bool = True,
    check_missing_fields: bool = False,
) -> Any:
    """
    Takes in the message type and a Python dictionary and returns a ROS message.

    :param message_type: The type name of the ROS message to return.
    :param dictionary: The values to set in the ROS message. The keys of the dictionary represent
        fields of the message.
    :param kind: Whether to create a message, service request or service response (valid values: "message",
        "request", "response").
    :param strict_mode: If strict_mode is set, an AttributeError will be raised if the input dictionary contains
        extra fields that are not present in the ROS message.
    :param check_missing_fields: If this parameter is set, a ValueError will be raised if the input dictionary
        is missing an entry for some field of the ROS message.
    :raises AttributeError: If the message does not have a field provided in the input dictionary.
    :raises TypeError: If a message value does not match its field type.
    :raises ValueError: If the input dictionary is incomplete (i.e., is missing an entry for some field
        of the ROS message) or if an invalid value was passed to `kind`.

    Example:
        >>> msg_type = "std_msgs/msg/String"
        >>> dict_msg = { "data": "Hello, Robot" }
        >>> convert_dictionary_to_ros_message(msg_type, dict_msg)
        std_msgs.msg.String(data='Hello, Robot')

        >>> msg_type = "std_srvs/srv/SetBool"
        >>> dict_msg = { "data": True }
        >>> kind = "request"
        >>> convert_dictionary_to_ros_message(msg_type, dict_msg, kind)
        std_srvs.srv.SetBool_Request(data=True)
    """
    if kind == 'message':
        message_class = get_message(message_type)
        message = message_class()
    elif kind == 'request':
        service_class = get_service(message_type)
        message = service_class.Request()
    elif kind == 'response':
        service_class = get_service(message_type)
        message = service_class.Response()
    else:
        raise ValueError('Unknown kind "%s".' % kind)

    set_message_fields(message, dictionary, strict_mode, check_missing_fields)
    return message


def set_message_fields(
    msg: Any, values: Dict[str, Any], strict_mode: bool = True, check_missing_fields: bool = False
) -> None:
    """
    Set the fields of a ROS message.

    This method was copied and modified from rosidl_runtime_py.set_message.set_message_fields .

    :param msg: The ROS message to populate.
    :param values: The values to set in the ROS message. The keys of the dictionary represent
        fields of the message.
    :param strict_mode: If strict_mode is set, an AttributeError will be raised if the input dictionary contains
        extra fields that are not present in the ROS message.
    :param check_missing_fields: If this parameter is set, a ValueError will be raised if the input dictionary
        is missing an entry for some field of the ROS message.
    :raises AttributeError: If the message does not have a field provided in the input dictionary.
    :raises TypeError: If a message value does not match its field type.
    :raises ValueError: If the input dictionary is incomplete (i.e., is missing an entry for some field
        of the ROS message).
    """
    if values is None:
        values = {}
    try:
        items = values.items()
    except AttributeError:
        raise TypeError("Value '%s' is expected to be a dictionary but is a %s" % (values, type(values).__name__))

    remaining_message_fields = dict(_get_message_fields(msg))

    for field_name, field_value in items:
        if field_value is None:
            continue
        try:
            field = getattr(msg, field_name)
            del remaining_message_fields[field_name]
        except AttributeError as e:
            if strict_mode:
                raise e
            else:
                continue
        field_type = type(field)
        if field_type is array.array:
            if isinstance(field_value, (str, bytes)):
                # If field_value is not properly base64 encoded and there are non-base64-alphabet characters in the
                # input, a binascii.Error will be raised.
                field_value = list(base64.b64decode(field_value, validate=True))
            value = field_type(field.typecode, field_value)
        elif field_type is np.ndarray:
            if isinstance(field_value, (str, bytes)):
                # If field_value is not properly base64 encoded and there are non-base64-alphabet characters in the
                # input, a binascii.Error will be raised.
                field_value = list(base64.b64decode(field_value, validate=True))
            value = np.array(field_value, dtype=field.dtype)
        elif type(field_value) is field_type:
            value = field_value
        else:
            try:
                value = field_type(field_value)
            except TypeError:
                value = field_type()
                set_message_fields(value, field_value, strict_mode, check_missing_fields)
        rosidl_type = get_message_slot_types(msg)[field_name]
        # Check if field is an array of ROS messages
        if isinstance(rosidl_type, AbstractNestedType):
            if isinstance(rosidl_type.value_type, NamespacedType):
                field_elem_type = import_message_from_namespaced_type(rosidl_type.value_type)
                for n in range(len(value)):
                    submsg = field_elem_type()
                    set_message_fields(submsg, value[n], strict_mode, check_missing_fields)
                    value[n] = submsg
        setattr(msg, field_name, value)

    if check_missing_fields and remaining_message_fields:
        error_message = 'fields in dictionary missing from ROS message: "{0}"'.format(
            list(remaining_message_fields.keys())
        )
        raise ValueError(error_message)


def convert_ros_message_to_dictionary(message: Any, base64_encoding: bool = True) -> OrderedDict:
    """
    Takes in a ROS message and returns an OrderedDict.

    Example:
        >>> import std_msgs.msg
        >>> ros_message = std_msgs.msg.UInt32(data=42)
        >>> convert_ros_message_to_dictionary(ros_message)
        OrderedDict([('data', 42)])
    """

    return message_to_ordereddict(message, base64_encoding=base64_encoding)


def message_to_ordereddict(
    msg: Any,
    *,
    base64_encoding: bool = True,
    truncate_length: int = None,
    no_arr: bool = False,
    no_str: bool = False,
) -> OrderedDict:
    """
    Convert a ROS message to an OrderedDict.

    This method was copied and modified from rosidl_runtime_py.convert.message_to_ordereddict.

    :param msg: The ROS message to convert.
    :param base64_encoding: If this parameter is true, encode all variable-size `uint8[]` or fixed-size `uint8[n]`
           fields using Base64 encoding. This saves a lot of space when converting large messages
           (such as sensor_msgs/Image) to json format.
           If this parameter is `False`, these fields will be converted to `array.array` (for variable-size fields) or
           `numpy.ndarray` (for fixed-size fields) instead.
    :param truncate_length: Truncate values for all message fields to this length.
        This does not truncate the list of fields (ie. the dictionary keys).
    :param no_arr: Exclude array fields of the message.
    :param no_str: Exclude string fields of the message.
    :returns: An OrderedDict where the keys are the ROS message fields and the values are
        set to the values of the input message.
    """
    d = OrderedDict()

    # We rely on __slots__ retaining the order of the fields in the .msg file.
    for field_name, field_type in zip(msg.__slots__, msg.SLOT_TYPES):
        value = getattr(msg, field_name, None)

        value = _convert_value(
            value,
            base64_encoding=base64_encoding,
            field_type=field_type,
            truncate_length=truncate_length,
            no_arr=no_arr,
            no_str=no_str,
        )
        # Remove leading underscore from field name
        d[field_name[1:]] = value
    return d


def _convert_value(
    value, *, base64_encoding: bool = True, field_type=None, truncate_length=None, no_arr=False, no_str=False
):
    if isinstance(value, bytes):
        if truncate_length is not None and len(value) > truncate_length:
            value = bytearray(''.join([chr(c) for c in value[:truncate_length]]) + '...', 'utf-8')
    elif isinstance(value, str):
        if no_str is True:
            value = '<string length: <{0}>>'.format(len(value))
        elif truncate_length is not None and len(value) > truncate_length:
            value = value[:truncate_length] + '...'
    elif isinstance(value, (list, tuple, array.array, np.ndarray)):
        # Since arrays and ndarrays can't contain mixed types convert to list
        typename = tuple if isinstance(value, tuple) else list

        if (
            base64_encoding
            and isinstance(field_type, (Array, UnboundedSequence))
            and type(field_type.value_type) is BasicType
            and field_type.value_type.typename == 'uint8'
        ):
            value = base64.b64encode(value).decode('utf-8')
        elif no_arr is True and field_type is not None:
            value = __abbreviate_array_info(value, field_type)
        elif truncate_length is not None and len(value) > truncate_length:
            # Truncate the sequence
            value = value[:truncate_length]
            # Truncate every item in the sequence
            value = typename(
                [
                    _convert_value(
                        v,
                        base64_encoding=base64_encoding,
                        truncate_length=truncate_length,
                        no_arr=no_arr,
                        no_str=no_str,
                    )
                    for v in value
                ]
                + ['...']
            )
        else:
            # Convert every item in the list
            value = typename(
                [
                    _convert_value(
                        v,
                        base64_encoding=base64_encoding,
                        truncate_length=truncate_length,
                        no_arr=no_arr,
                        no_str=no_str,
                    )
                    for v in value
                ]
            )
    elif isinstance(value, dict) or isinstance(value, OrderedDict):
        # Convert each key and value in the mapping
        new_value = {} if isinstance(value, dict) else OrderedDict()
        for k, v in value.items():
            # Don't truncate keys because that could result in key collisions and data loss
            new_value[_convert_value(k)] = _convert_value(
                v,
                base64_encoding=base64_encoding,
                truncate_length=truncate_length,
                no_arr=no_arr,
                no_str=no_str,
            )
        value = new_value
    elif isinstance(value, np.number):
        value = value.item()
    elif not isinstance(value, (bool, float, int)):
        # Assuming value is a message since it is neither a collection nor a primitive type
        value = message_to_ordereddict(
            value,
            base64_encoding=base64_encoding,
            truncate_length=truncate_length,
            no_arr=no_arr,
            no_str=no_str,
        )
    return value


def __abbreviate_array_info(value, field_type):
    value_type_name = __get_type_name(field_type.value_type)
    if isinstance(field_type, rosidl_parser.definition.Array):
        return '<array type: {0}[{1}]>'.format(value_type_name, field_type.size)
    elif isinstance(field_type, rosidl_parser.definition.BoundedSequence):
        return '<sequence type: {0}[{1}], length: {2}>'.format(value_type_name, field_type.maximum_size, len(value))
    elif isinstance(field_type, rosidl_parser.definition.UnboundedSequence):
        return '<sequence type: {0}, length: {1}>'.format(value_type_name, len(value))
    return 'unknown'


def __get_type_name(value_type):
    if isinstance(value_type, rosidl_parser.definition.BasicType):
        return value_type.typename
    elif isinstance(value_type, rosidl_parser.definition.AbstractString):
        return 'string'
    elif isinstance(value_type, rosidl_parser.definition.AbstractWString):
        return 'wstring'
    elif isinstance(value_type, rosidl_parser.definition.NamedType):
        return value_type.name
    elif isinstance(value_type, rosidl_parser.definition.NamespacedType):
        return '/'.join(value_type.namespaced_name())
    else:
        return 'unknown'


def _get_message_fields(message):
    # Workaround due to new python API https://github.com/ros2/rosidl_python/issues/99
    slots = [slot[1:] for slot in message.__slots__]  # remove leading underscore
    return zip(slots, message.SLOT_TYPES)


if __name__ == "__main__":
    import doctest

    doctest.testmod()
